Cryptodoc Update for ML-DSA (#243)
sehlen-bsi/botan-docs: docs/cryptodoc/src/05_08_ml_dsa.rst

.. _pubkey/ml_dsa:

ML-DSA
======

Botan implements the Module-Lattice-based Digital Signature Algorithm (ML-DSA)
in :srcref:`src/lib/pubkey/dilithium`. The implementation is based on

> [FIPS-204]_. The list of supported algorithm parametres is listed in the table

Reviewer: ```suggestion
[FIPS-204]_. The list of supported algorithm parameters is listed in the table
```

:ref:`pubkey/ml_dsa/params`.

.. _pubkey/ml_dsa/params:

.. table:: Supported ML-DSA Parameter Sets (see Table 1 of [FIPS-204]_)

   +------------+-----------+-----------+---------------+-----------+--------------+-----------------+------------------+------------------+---------------+--------------+---------------+----------------+
   | Mode       | :math:`q` | :math:`n` | :math:`\zeta` | :math:`d` | :math:`\tau` | :math:`\lambda` | :math:`\gamma_1` | :math:`\gamma_2` | :math:`(k,l)` | :math:`\eta` | :math:`\beta` | :math:`\omega` |

>    +============+===========+===========+===============+===========+==============+=================+==================+==================+===============+==============+===============+================+

Reviewer: Maybe remove $n$? It does not appear in the table and of FIPS-204 and is also never referenced in this chapter.

Author: Yeah, I went back and forth with that one, too. Will remove it.

   | ML-DSA-8x7 | 8380417   | 256       | 1753          | 13        | 60           | 256             | 2\ :sup:`19`     | :math:`(q-1)/32` | (8,7)         | 2            | 120           | 75             |
   +------------+-----------+-----------+---------------+-----------+--------------+-----------------+------------------+------------------+---------------+--------------+---------------+----------------+

The parameter sets claim a NIST security level of 2, 3, and 5 respectively.

.. _pubkey/ml_dsa/internals:

Algorithm Internals
-------------------

[FIPS-204]_ describes three primary operations: key generation, signing, and
verification.

Internally, those operations are further split into two functional layers:
the public-facing ML-DSA (Section 5), and ML-DSA-internal (Section 6). ML-DSA
and ML-DSA-internal decouple the actual high-level logic from the generation of
randomness. The functions in ML-DSA-internal receive pre-determined random bytes
as needed and are, therefore, fully deterministic.

ML-DSA is a Schnorr-like signature where the typically interactive protocol is
made non-interactive by pseudorandomly deriving the verifier's challenge from

> the commitment and the message to be signed.

Author: Left one sentence, more as a transition to the table. Frankly, I had hoped that this table is somewhat self-explanatory.

.. _pubkey/ml_dsa/components:

.. table:: ML-DSA Components and File Locations

   +-------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------------------------------------+--------------------+
   | Component                                                                                                                                             | Purpose                                                        | [FIPS-204]_        |
   +=======================================================================================================================================================+================================================================+====================+

>    | :ref:`Types <pubkey/ml_dsa/types>` (:srcref:`src <[src/lib/pubkey/dilithium/dilithium_common]/dilithium_types.h>`)                                    | Strong Types                                                   | n/a                |

Reviewer: Personally, I like this variant to link to GitHub. Maybe we can discuss in our next JF if we want to adapt this to all overview lists.

Author: Yeah, simply using `(src)` really is a crutch here to safe space in the rendering. `.../dilithium_symmetric_primitives.h` really is too wide for this table. ;) I like the idea of using a symbol here, perhaps even integrate it as a feature in `:srcref:`. Maybe `:srcref-icon:`, or perhaps as a special case when the entire URL is marked as ellipsis using brackets, like: ``:srcref:'[src/lib/pubkey/dilithium/dilithium_common/dilithium_types.h]'``

Author: Please open a ticket for this.

.. _pubkey/ml_dsa/types:

Strong Types
^^^^^^^^^^^^

ML-DSA uses strong types and type aliases to represent the various value
types involved in the algorithm. This approach binds the semantic meaning of
values to their types, resulting in a more robust interface and self-documenting
code. Type aliases are defined for ML-DSA polynomials, polynomial vectors, and
polynomial matrices, as well as their NTT representations. All bitstrings,
including various hash values, random seeds, and others, are encapsulated as
strong types.

.. _pubkey/ml_dsa/constants:

Paramter Instantiations
^^^^^^^^^^^^^^^^^^^^^^^

Botan's ``DilithiumConstants`` class contains all parameters and constants
outlined in Section 4 of [FIPS-204]_ (see :ref:`pubkey/ml_dsa/params`).
Additionally, the class contains parameters implicitly derived from these
constants, such as key and ciphertext sizes, along with various intermediate
value sizes required within internal algorithms.

Also, this class contains the theoretical XOF-bounds outlined in Appendix C of

> [FIPS-204]_ used as a guard rail for the various rejection sampling operations

Reviewer: ```suggestion
[FIPS-204]_ used as a guardrail for the various rejection sampling operations
```

> within the ML-DSA implementation.

Author: # 🥺

> within the ML-DSA implementation.

Author: ... love the level of nit-marking, though.

.. _pubkey/ml_dsa/polynomials:

Polynomial Operations
^^^^^^^^^^^^^^^^^^^^^

ML-DSA relies extensively on polynomials within the polynomial ring :math:`R_q`,
utilizing vectors and matrices of polynomials, both inside and outside the NTT
domain. Botan uses :ref:`strong types <pubkey/ml_dsa/types>` to distinguish
polynomials and polynomial vectors as ``DilithiumPoly`` and
``DilithiumPolyVec``, as well as their NTT counterparts ``DilithiumPolyNTT`` and
``DilithiumPolyVecNTT``. Matrices only appear in the NTT domain and are
represented by the class ``DilithiumPolyMatNTT``.

> ML-KEM, as defined in [FIPS-202]_, also employs polynomials, leading to shared

Author: # 🦅 👀

support the primary algorithm. Table :ref:`pubkey/ml_dsa/algos` offers a summary
of those functions that are exposed to the rest of the implementation. The
:srcref:`[src/lib/pubkey/dilithium/dilithium_common]/dilithium_algos.cpp`
contains additional functions that are used within this module only, such as the
encoding functionality from [FIPS-204]_ Section 7.1.

.. _pubkey/ml_dsa/algos:

.. table:: ML-DSA Algorithms Overview

   +---------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+-------------+
   | Function                                                                                                                              | Description                                                                                | [FIPS-204]_ |
   +=======================================================================================================================================+============================================================================================+=============+

>    | :srcref:`encode_public_key <[src/lib/pubkey/dilithium/dilithium_common]/dilithium_algos.cpp:327|encode_public_key>`                   | Byte encoding of a public key                                                              | 7.2 A. 22   |

Author: Algorithm is enough, despite itching "but now that we have it"-vibes.

   | :srcref:`use_hint <[src/lib/pubkey/dilithium/dilithium_common]/dilithium_algos.cpp:918|use_hint>`                                     | Lets the verifier decompress the signature                                                 | 7.4 A. 40   |
   +---------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+-------------+
   | :srcref:`infinity_norm_within_bound <[src/lib/pubkey/dilithium/dilithium_common]/dilithium_algos.cpp:936|infinity_norm_within_bound>` | Given vector :math:`v` and :math:`bound`, validates that :math:`\|v\|_{\infty} \geq bound` | n/a         |
   +---------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+-------------+

> .. [#dilithium_comp] The private key encoding and decoding functions are used

Author: This bugged me too, and then I decided that I can live with it.

   for the legacy support of Dilithium (round 3) only. Botan's ML-DSA

>    implementation exclusively store its private keys as the secret seed

Reviewer: ```suggestion
   implementation exclusively stores its private keys as the secret seed
```

be found in :srcref:`[src/lib/pubkey/dilithium]/ml_dsa/ml_dsa_impl.h` and
:srcref:`[src/lib/pubkey/dilithium/dilithium_common/dilithium_shake]/dilithium_shake_xof.h`.

.. _pubkey/ml_dsa/keys:

Internal Key Representation
^^^^^^^^^^^^^^^^^^^^^^^^^^^

The :srcref:`Dilithium_PublicKeyInternal
<src/lib/pubkey/dilithium/dilithium_common/dilithium_keys.h:33|Dilithium_PublicKeyInternal>`
and :srcref:`Dilithium_PrivateKeyInternal
<src/lib/pubkey/dilithium/dilithium_common/dilithium_keys.h:68|Dilithium_PrivateKeyInternal>`
classes are the internal representation of the ML-DSA key pair in expandedd form.

Additionally, the :srcref:`Dilithium_Keypair_Codec
<src/lib/pubkey/dilithium/dilithium_common/dilithium_keys.h:23|Dilithium_Keypair_Codec>`
serves as a customization point for the key encoding and decoding functions that
differ between ML-DSA (:math:`\xi` only) and Dilithium (round 3) (partially
expanded key format as specified in [FIPS-204]_). By *always* expanding the
private key from the secret seed :math:`\xi`, sanity checks during decoding of
the key pair can be omitted.

.. _pubkey/ml_dsa/ml_dsa_impl:

ML-DSA Specifics
^^^^^^^^^^^^^^^^

> This module provides concrete ML-DSA specific implementations for the

Reviewer: ```suggestion
This module provides concrete ML-DSA-specific implementations for the
```

customization points outlined in :ref:`pubkey/ml_dsa/primitives` and
:ref:`pubkey/ml_dsa/keys`. Namely:

  * :srcref:`ML_DSA_Expanding_Keypair_Codec <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:21|ML_DSA_Expanding_Keypair_Codec>`
    Implements encoding and decoding of ML-DSA private keys by serializing the
    private seed :math:`\xi` and/or expanding the deserialized seed into the
    private key representation outlined in :ref:`pubkey/ml_dsa/keys`.
  * :srcref:`ML_DSA_MessageHash <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:28|ML_DSA_MessageHash>`
    Implements the transformation of the user-provided message :math:`M` into
    the message representation :math:`\mu`. This includes the incorporation of
    the domain separations outlined in [FIPS-204]_ Sectoin 5.2 Algorithm 2.
  * :srcref:`ML_DSA_Symmetric_Primitives
    <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:54|ML_DSA_Symmetric_Primitives>`
    Implements the ML-DSA specific symmetric primitives based on the specified
    SHAKE-based XOFs, the optional hedged randomization of :math:`H(K \| rnd \| \mu)`,

>     and the domain separator for expanding :math:`\rho`, :math:`\rho'`, and

Author: This would need several reference. Not sure that this is any helpful, it'll disturb the readability for sure, though.

    :math:`K` from the private seed :math:`\xi`.

.. _pubkey/ml_dsa/ml_dsa_api:

Public API
^^^^^^^^^^

The :srcref:`Dilithium_PublicKey
<src/lib/pubkey/dilithium/dilithium_common/dilithium.h:67|Dilithium_PublicKey>`
and :srcref:`Dilithium_PrivateKey
<src/lib/pubkey/dilithium/dilithium_common/dilithium.h:115|Dilithium_PrivateKey>`
classes serve as Botan's public API for public and private ML-DSA keys,
respectively. The :srcref:`DilithiumMode
<src/lib/pubkey/dilithium/dilithium_common/dilithium.h:21|DilithiumMode>` class
is used to select the desired parameter set.

New applications that do not rely on the pre-standard Dilithium round 3
implementations are strongly advised to use the type aliases for ML-DSA defined
in :srcref:`[src/lib/pubkey/dilithium/ml_dsa]/ml_dsa.h`.

.. _pubkey/ml_dsa/kyber_compat:

Dilithium Compatibility
-----------------------

The final ML-DSA standard is not compatible with the round 3 submission of

> Dilithium. Botan did provide support for Dilithium and Dilithium-AES as

Reviewer: ```suggestion
Dilithium. Botan has provided support for Dilithium and Dilithium-AES as
```
(says Grammarly)

Author: I think it is right. "did provide" would be correct, if the support was removed in the past.

specified in [Dilithium-R3]_ since April 2023. This support is still available

> and can be activated by enabling the ``dilithium`` or ``dilithium_aes`` modules.

Author: Fair.

> Note that Dilithium-AES is already deprecated, and both Dilithium and

Reviewer: ```suggestion
Note that Dilithium-AES has already been deprecated, and both Dilithium and
```

<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:403|Dilithium_PrivateKey>`.
This mostly delegates the actual key generation to the internal function
:srcref:`expand_keypair
<src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:665|expand_keypair>`
that follows [FIPS-204]_ Section 6.1 Algorithm 6.

.. admonition:: Dilithium_PrivateKey::Dilithium_PrivateKey / Dilithium_Algos::expand_keypair

   **Input:**

   - ``rng``: random number generator
   - ``mode``: ML-DSA parameter set descriptor

   **Output:**

   - ``sk``: private signing key
   - ``pk``: public verification key

   **Steps:**

   1. Generate a random 32-byte seed :math:`\xi` using ``rng``
   2. :math:`(\rho, \rho', K) = H(\xi)` (32, 64, and 32 bytes respectively)
   3. Sample matrix :math:`\hat{A}` from :math:`\rho` using ``expand_A``
   4. Sample vectors :math:`s_1` and :math:`s_2` from :math:`\rho'` using ``expand_s``
   5. Calculate :math:`(t_1, t_0)` from :math:`\hat{A}`, :math:`s_1`, and :math:`s_2` using ``compute_t1_and_t0`` (see :srcref:`here <src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:310|compute_t1_and_t0>`)

>       1. :math:`t = NTT^{-1}(\hat{A} \cdot NTT(s_1)) + s_2`

Author: Yeah, probably better. I'm guessing you did the same in ML-KEM, no?

   6. :math:`pk = (\rho, t_1)` and :math:`sk = (\xi, K, s_1, s_2, t_0)`

   **Notes:**

   - Step 1 corresponds to [FIPS-204]_ Algorithm 1
   - Steps 2-5 correspond to [FIPS-204]_ Algorithm 6
   - Step 6 returns the key pair in Botan's internal representation. The
     encoding and hashing of the encoded public key are done later and on
     demand.

.. _pubkey/ml_dsa/signing:

Signing
-------

Signature generation as specified in [FIPS-204]_ Algorithms 2 and 7 are
implemented in :srcref:`Dilithium_Signature_Operation::sign
<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:153|sign>` with the
preparation of the message representative :math:`\mu` being done in
:srcref:`DilithiumMessageHash
<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.

.. admonition:: Dilithium_Signature_Operation::sign

   **Input:**

   - ``sk``: private signing key, with :math:`\hat{A}` and :math:`\hat{s_1}`, :math:`\hat{s_2}`, and :math:`\hat{t_0}` in NTT domain
   - ``M``: message to be signed

   **Output:**

   - ``signature``: the valid signature

   **Steps:**

>    1. Calculate the message representative :math:`\mu = H(sk.tr \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)

Reviewer 2: I guess instead of `sk.tr`this should be `H(pk)`or similar since the hashing is done on demand as stated in the keygen? Or is it a member of `sk` that is computed on first use?

Author: The public key is not an input variable for signing. And the private key is defined to contain `tr = H(pk)`. But you're right, this is not obvious, I'll note that `sk.tr = H(pk)` in the `Input` section.

>    1. Calculate the message representative :math:`\mu = H(sk.tr \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)

Author: Are you sure? I've seen a similar notation in SLH-DSA as well. For instance:

https://github.com/sehlen-bsi/botan-docs/blob/e2ab5e3c59c2a7d01e75fab00e2ed1b8733dd91a/docs/cryptodoc/src/05_07_slh_dsa.rst?plain=1#L422

>    1. Calculate the message representative :math:`\mu = H(sk.tr \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)

Reviewer 2: True, must have mixed that up. nvm then, nothing to see here.

   2. :math:`\rho' = H(sk.K \| rnd \| \mu)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:68|H_maybe_randomized>`)
   3. Run the rejection sampling loop (incrementing the nonce :math:`\kappa` by :math:`l` in each iteration)

      1. Expand :math:`y` from :math:`\rho'` and :math:`\kappa` using ``expand_mask``
      2. :math:`\hat{w} = \hat{A} \cdot NTT(y)`
      3. :math:`w = NTT^{-1}(\hat{w})`
      4. :math:`(w_1, w_0) = decompose(w)`
      5. :math:`\tilde{c} = H(\mu \| w_1)` (:math:`w_1` is encoded using ``encode_commitment``)
      6. :math:`\hat{c} = NTT(sample\_in\_ball(\tilde{c}))`
      7. :math:`cs_1 = NTT^{-1}(\hat{c} \cdot sk.\hat{s_1})`
      8. :math:`z = y + cs_1`

>       9. *Retry* iff :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)

Reviewer 2: ```suggestion
      9. *Retry* if :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)
```
Similar for whole enumeration

      10. :math:`cs_2 = NTT^{-1}(\hat{c} \cdot sk.\hat{s_2})`
      11. :math:`r_0 = w_0 - cs_2`

>       12. *Retry* iff :math:`\|r_0\|_{\infty} \geq \gamma_2 - \beta` (see ``infinity_norm_within_bound``))

Reviewer 2: In [FIPS-204] there is only one check. Does that matter?

Author: I don't see why it would matter. When you combine the two conditions into one if-statement combined with an "or", the program flow would exit if the first condition is true. If you wanted to avoid this side-channel entirely, you would have to take extra measures regardless.

In our case, this distinction is easier to measure, because we do some extra calculation between the conditions. But I would argue that, if this were a problem, the spec would also emphasize that.

Also, the reference implementation splits those conditions similarly to us:

https://github.com/pq-crystals/dilithium/blob/cbcd8753a43402885c90343cd6335fb54712cda1/ref/sign.c#L161-L188

      13. :math:`ct_0 = NTT^{-1}(\hat{c} \cdot sk.\hat{t_0})`

>       14. *Retry* iff :math:`\|c\|_{\infty} \geq \gamma_2` (see ``infinity_norm_within_bound``)

Author: Good catch!

      15. :math:`h = make\_hint(r_0 + ct_0, w_1)`
      16. *Retry* if the Hamming weight of :math:`h > \omega`

   4. :math:`\sigma = (\tilde{c}, z, h)` encoded using ``encode_signature``

   **Notes:**

>    - This algorithm description assumes that the private signing key is expanded
>      into the internal representation already. Additionally, the expansion of

Reviewer: ```suggestion
   - This algorithm description assumes that the private signing key has already been expanded
     into the internal representation. Additionally, the expansion of
```

     :math:`\hat{A}`, as well as the NTT for :math:`s_1`, :math:`s_2`, and
     :math:`t_0` are done :srcref:`prior to the actual signing operation
     <src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:131|Dilithium_Signature_Operation>`

>      to armortize the complexity of these operations across multiple consecutive

Reviewer: ```suggestion
     to amortize the complexity of these operations across multiple consecutive
```

     signature generations.
   - Step 1: Botan 3.6.0 does not yet support the application-defined context
     string as specified in [FIPS-204]_ Algorithm 2. See `GitHub #4376
     <https://github.com/randombit/botan/issues/4376>`_.
   - Step 3.12: We calculate :math:`r_0 = w_0 - cs_2`, this is equivalent to
     :math:`r_0 = LowBits(w - cs_2)` as specified in [FIPS-204]_ Algorithm 7.

>    - Step 3.15: We generate the hint from :math:`w_0 - cs_2 + ct_0` this is equivalent

Reviewer 2: ```suggestion
   - Step 3.15: We generate the hint from :math:`w_0 - cs_2 + ct_0`. This is equivalent
```

>      to the specification's :math:`w - cs_2 + ct_0`.

Author: I removes the detailed calculation but otherwise used your suggestion.

.. _pubkey/ml_dsa/verification:

Signature Verification
----------------------

Signature verification as specified in [FIPS-204]_ Algorithms 3 and 8 are
implemented in :srcref:`Dilithium_Verification_Operation::is_valid_signature
<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:269|is_valid_signature>`
with the preparation of the message representative :math:`\mu` being done in
:srcref:`DilithiumMessageHash
<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.

.. admonition:: Dilithium_Verification_Operation::is_valid_signature

   **Input:**

   - ``pk``: public verification key, with :math:`\hat{A}` and :math:`\hat{t_1'} = NTT(t_1 \cdot 2^{d})` in NTT domain
   - ``M``: message to be verified
   - ``signature``: the signature to be verified

   **Output:**

   - ``ok``: boolean value whether or not the signature is valid

   **Steps:**

>    1. Calculate the message representative :math:`\mu = H(H(pk) \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)

Reviewer 2: Similar to Sign: Consistency with SLH-DSA

   2. Decode the signature into :math:`(\tilde{c}, z, h)` using ``decode_signature``

>    3. *Abort with "not valid"* if the Hamming weight of :math:`h > \omega`

Reviewer 2: I don't see this condition in the standard?

Author: This is part of [the Dilithium round 3 spec](https://pq-crystals.org/dilithium/data/dilithium-specification-round3.pdf) (Section 3, Figure 4, line 31):

![image](https://github.com/user-attachments/assets/6fff9259-e9e8-4bbe-a452-f711d37d519c)

I'm guessing that it was removed during standardization. My guess is that it would reject signatures with hints revealing more than they should (?). But that's not the immediate problem of the verifier...

Oh, and I just saw that there's a remnant of this check in FIPS-204 Section 6.3 "ML-DSA Verifying (Internal)", where it reads in the last paragraph of the textual description:

> [...], the verifier checks that all of the coefficients of z are sufficiently small, **h contains no more than 𝜔 nonzero coefficients**, and c matches the hash c' [...]

In Botan, we have [a TODO that this check is not part of the standard anymore](https://github.com/randombit/botan/blob/dc326edf1204ec0875f807ffc475717d4c96410f/src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp#L286-L290). But then, it actually kind of is still in the standard.

I'll add a Note describing the situation.

Author: Side note: the hint packing algorithms (FIPS-204 A. 20 + 21) are designed in a way that they cannot store a hint that has a hamming weight greater than 𝜔. Thus, using the standardized signature format, it isn't possible to create a signature that would ever fail this check. We should remove it from the implementation.

>    4. *Abort with "not valid"* if :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)

Reviewer 2: Is the order of aborting relevant (SCA)?

Author: During validation that is probably less of a problem, unless perhaps, when we're dealing with confidential signatures. FWIW: the reference implementation performs this check before any other calculations as well: 

https://github.com/pq-crystals/dilithium/blob/cbcd8753a43402885c90343cd6335fb54712cda1/ref/sign.c#L270-L316

   5. :math:`\hat{c} = NTT(sample\_in\_ball(\tilde{c}))`

>    6. :math:`w'_{approx} = \hat{A} \cdot NTT(z) - \hat{c} \cdot \hat{t_1'}`

Reviewer: The spec names it with a capital A:
```suggestion
   6. :math:`w'_{Approx} = \hat{A} \cdot NTT(z) - \hat{c} \cdot \hat{t_1'}`
```

   7. :math:`w_1' = use\_hint(w'_{approx}, h)`

>    8. :math:`\tilde{c'} = H(\mu, w_1')` (:math:`w_1'` is encoded using ``encode_commitment``)
>    9. Iff :math:`\tilde{c} = \tilde{c'}` *return "valid"*, else *"not valid"*

Reviewer: ```suggestion
   8. :math:`\tilde{c}' = H(\mu, w_1')` (:math:`w_1'` is encoded using ``encode_commitment``)
   9. If :math:`\tilde{c} = \tilde{c}'` *return "valid"*, else *"not valid"*
```
Looks more like in the spec. Also the _iff_ is not needed, I think, because of the _else_.

   **Notes:**

   - This algorithm description assumes that the public verification key is
     deserialized into the internal representation already. Additionally, the

>      expansion of :math:`\hat{A}`, as well as the prepartion of
>      :math:`\hat{t_1'} = NTT(t_1 \cdot 2^{d})` are done :srcref:`prior to the actual
>      verification operation
>      <src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:251|Dilithium_Verification_Operation>`
>      to armortize the complexity of these operations across multiple consecutive

Reviewer: ```suggestion
     expansion of :math:`\hat{A}`, as well as the preparation of
     :math:`\hat{t_1'} = NTT(t_1 \cdot 2^{d})` are done :srcref:`prior to the actual
     verification operation
     <src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:251|Dilithium_Verification_Operation>`
     to amortize the complexity of these operations across multiple consecutive
```

Dilithium. Botan has provided support for Dilithium and Dilithium-AES as
specified in [Dilithium-R3]_ since April 2023. This support is still available,
can be activated by enabling the ``dilithium`` or ``dilithium_aes`` modules,
and can be used via a ``DilithiumMode`` parameter set.

Note that Dilithium-AES has already been deprecated, and both Dilithium and
Dilithium-AES may be removed as early as the next major release of the library.
It is not advisable to use any other variant than the ones specified in
[FIPS-204]_.

.. _pubkey/ml_dsa/keygen:

Key Generation
--------------

Generating a fresh ML-DSA key pair as specified in [FIPS-204]_ Section 5.1
Algorithm 1, is available in the constructor of :srcref:`Dilithium_PrivateKey
<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:403|Dilithium_PrivateKey>`.
This mostly delegates the actual key generation to the internal function
:srcref:`expand_keypair
<src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:665|expand_keypair>`
that follows [FIPS-204]_ Section 6.1 Algorithm 6.

.. admonition:: Dilithium_PrivateKey::Dilithium_PrivateKey / Dilithium_Algos::expand_keypair

   **Input:**

>    - :math:``rng``: random number generator

Reviewer 3: Somehow I still see the ":math:" in the PDF.

Reviewer 3: I think it's due to the double backtickts here and for mode, sk and pk.
But maybe :math: is not needed here?

Reviewer 2: agreed, removed all unnecessary and incorrectly rendered `:math:`s

   - :math:``mode``: ML-DSA parameter set descriptor

   **Output:**

   - :math:``sk``: private signing key
   - :math:``pk``: public verification key

   **Steps:**

   1. Generate a random 32-byte seed :math:`\xi` using ``rng``
   2. :math:`(\rho, \rho', K) = H(\xi)` (32, 64, and 32 bytes respectively)
   3. Sample matrix :math:`\hat{A}` from :math:`\rho` using ``expand_A``
   4. Sample vectors :math:`s_1` and :math:`s_2` from :math:`\rho'` using ``expand_s``
   5. Calculate :math:`(t_1, t_0)` from :math:`\hat{A}`, :math:`s_1`, and :math:`s_2` using ``compute_t1_and_t0`` (see :srcref:`here <src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:310|compute_t1_and_t0>`)

>       1. :math:`t = NTT^{-1}(\hat{A} \circ NTT(s_1)) + s_2`

Reviewer 3: The notation here is different than in ML-KEM. ML-KEM uses ``ntt`` and ``ntt_inverse``. This should be unified.

Reviewer 2: changed

      2. :math:`(t_1, t_0) = power2round(t)` (see :srcref:`here <src/lib/pubkey/dilithium/dilithium_common/dilithium_algos.cpp:746|power2round>`)

   6. :math:`pk = (\rho, t_1)` and :math:`sk = (\xi, K, s_1, s_2, t_0)`

   **Notes:**

   - Step 1 corresponds to [FIPS-204]_ Algorithm 1
   - Steps 2-5 correspond to [FIPS-204]_ Algorithm 6
   - Step 6 returns the key pair in Botan's internal representation. The
     encoding and hashing of the encoded public key are done later and on
     demand.

.. _pubkey/ml_dsa/signing:

Signing
-------

Signature generation as specified in [FIPS-204]_ Algorithms 2 and 7 are
implemented in :srcref:`Dilithium_Signature_Operation::sign
<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:153|sign>` with the
preparation of the message representative :math:`\mu` being done in
:srcref:`DilithiumMessageHash
<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.

.. admonition:: Dilithium_Signature_Operation::sign

   **Input:**

>    - :math:``sk``: private signing key, with :math:`\hat{A}` and :math:`\hat{s_1}`, :math:`\hat{s_2}`, :math:`\hat{t_0}` in NTT domain, :math:`tr = H(pk)`, and :math:`K`

Reviewer 3: Same problem here.

Reviewer 3: Also, it is not explained, as far as I can see, that sk has the functionality to return Ahat and tr above (step 6 does not show this at least) but within the signing routine your accessing sk.tr etc.

Reviewer 2: IMO it is clear that `sk.<i>` refers to the tuple element `<i>` of `sk`. But we can make this more explicit if you have strong feelings about it.

Reviewer 2: Yes, but the tuple `sk` is defined in the input as also providing `tr = H(pk)`, i.e., this is pre-computed when loading `sk`. If this is confusing we can add a clarifying note,

Reviewer 2: I added a note explaining this

      2. :math:`\hat{w} = sk.\hat{A} \circ NTT(y)`
      3. :math:`w = NTT^{-1}(\hat{w})`
      4. :math:`(w_1, w_0) = decompose(w)`
      5. :math:`\tilde{c} = H(\mu \| w_1)` (:math:`w_1` is encoded using ``encode_commitment``)
      6. :math:`\hat{c} = NTT(sample\_in\_ball(\tilde{c}))`
      7. :math:`cs_1 = NTT^{-1}(\hat{c} \circ sk.\hat{s_1})`
      8. :math:`z = y + cs_1`
      9. *Retry* if :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)
      10. :math:`cs_2 = NTT^{-1}(\hat{c} \circ sk.\hat{s_2})`
      11. :math:`r_0 = w_0 - cs_2`
      12. *Retry* if :math:`\|r_0\|_{\infty} \geq \gamma_2 - \beta` (see ``infinity_norm_within_bound``))
      13. :math:`ct_0 = NTT^{-1}(\hat{c} \circ sk.\hat{t_0})`
      14. *Retry* if :math:`\|ct_0\|_{\infty} \geq \gamma_2` (see ``infinity_norm_within_bound``)
      15. :math:`h = make\_hint(r_0 + ct_0, w_1)`
      16. *Retry* if the Hamming weight of :math:`h > \omega`

   4. :math:`\sigma = (\tilde{c}, z, h)` encoded using ``encode_signature``

   **Notes:**

   - This algorithm description assumes that the private signing key has already

>      been expanded into the internal representation already. Additionally, the

Reviewer 3: ```suggestion
     been expanded into the internal representation. Additionally, the
```
Appears twice in the sentence.

Reviewer 3: Maybe the first note should come before the description and introduce the notation sk.(...) to obtain Ahat and tr etc

     expansion of ``A_hat``, as well as the NTT for ``s_1``,
     ``s_2``, and ``t_0`` are done :srcref:`prior to the actual signing
     operation
     <src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:131|Dilithium_Signature_Operation>`
     to amortize the complexity of these operations across multiple consecutive

>      signature generations. The same applies to calculation of ``tr``.

Reviewer: ```suggestion
     signature generations. The same applies to the calculation of ``tr``.
```

   - Step 1: Botan 3.6.0 does not yet support the application-defined context
     string as specified in [FIPS-204]_ Algorithm 2. See `GitHub #4376
     <https://github.com/randombit/botan/issues/4376>`_.
   - Steps 3.12, 3.15: Botan uses an optimization for hint generation as
     provided by the reference implementation. Instead of computing the hint
     based on :math:`(w - cs_2 + ct_0, -ct_0)`, Botan computes it using the
     inputs :math:`(w_0 - cs_2 + ct_0, w_1)`. Both computations are equivalent.

.. _pubkey/ml_dsa/verification:

Signature Verification
----------------------

Signature verification as specified in [FIPS-204]_ Algorithms 3 and 8 is
implemented in :srcref:`Dilithium_Verification_Operation::is_valid_signature
<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:269|is_valid_signature>`
with the preparation of the message representative :math:`\mu` being done in
:srcref:`DilithiumMessageHash
<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.

.. admonition:: Dilithium_Verification_Operation::is_valid_signature

   **Input:**

>    - :math:``pk``: public verification key, with :math:`\hat{A}` and :math:`\hat{t_1'} = NTT(t_1 \circ 2^{d})` in NTT domain

Reviewer 3: same problem here as above

Reviewer 2: see above

   - :math:``M``: message to be verified
   - :math:``signature``: the signature to be verified

   **Output:**

   - :math:``ok``: boolean value whether or not the signature is valid

   **Steps:**

   1. Calculate the message representative :math:`\mu = H(H(pk) \| 0x00 \| 0x00 \| M)` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)
   2. Decode the signature into :math:`(\tilde{c}, z, h)` using ``decode_signature``
   3. *Abort with "not valid"* if the Hamming weight of :math:`h > \omega`
   4. *Abort with "not valid"* if :math:`\|z\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)
   5. :math:`\hat{c} = NTT(sample\_in\_ball(\tilde{c}))`
   6. :math:`w'_{Approx} = \hat{A} \circ NTT(z) - \hat{c} \circ \hat{t_1'}`

>    7. :math:`w_1' = use\_hint(w'_{Approx}, h)`

Reviewer 3: Is use_hint a function name that should be enclosed in double backticks?

Reviewer 2: Yes, but double backticks within math notation interferes. We noticed that we only use math notation in ML-DSA algorithms, so I changed all the algorithms to double backticks for consistency with the other algorithms.

Signature verification as specified in [FIPS-204]_ Algorithms 3 and 8 is
implemented in :srcref:`Dilithium_Verification_Operation::is_valid_signature
<src/lib/pubkey/dilithium/dilithium_common/dilithium.cpp:269|is_valid_signature>`
with the preparation of the message representative ``mu`` being done in
:srcref:`DilithiumMessageHash
<src/lib/pubkey/dilithium/dilithium_common/dilithium_symmetric_primitives.h:31|DilithiumMessageHash>`.

.. admonition:: Dilithium_Verification_Operation::is_valid_signature

   **Input:**

>    - ``pk``: public verification key, with ``A_hat`` and ``t_1'_hat = ntt(t_1 * 2^d)`` in NTT domain

Reviewer: `t_1_'_hat` looks a bit weird. Maybe: `t_1_hat'` instead? Or is this different?

   - ``M``: message to be verified
   - ``signature``: the signature to be verified

   **Output:**

>    - ``ok``: Boolean value whether or not the signature is valid

Reviewer: All other parameter descriptions begin with lower letters.

Reviewer 2: I was taught this should be capitalized because it refers to George Boole, but after googling it seems like this case should not be capitalized. I'll revert...

Reviewer: Wow, I never noticed that boolean comes from George Boole 😄

   **Steps:**

   1. Calculate the message representative ``mu = H(H(pk) | 0x00 | 0x00 | M)`` (see :srcref:`here <src/lib/pubkey/dilithium/ml_dsa/ml_dsa_impl.h:36|start>`)
   2. Decode the signature into ``(c_tilde, z, h)`` using ``decode_signature``

>    3. *Abort with "not valid"* if the Hamming weight of ``h`` is greater than :math:`\omega`

Reviewer: Do we want to write `''omega''` instead of `:math:'\omega'`?

Reviewer 2: No, because omega is part of the parameter set

   4. *Abort with "not valid"* if :math:`\|` ``z`` :math:`\|_{\infty} \geq \gamma_1 - \beta` (see ``infinity_norm_within_bound``)
   5. ``c_hat = ntt(sample_in_ball(c_tilde))``

>    6. ``w'_approx = A_hat * ntt(z) - c_hat * t_1'_hat``

Reviewer: Alternatively: `w_approx'`
